src/hnsw/product_quantization.c

```diff
@@ -0,0 +1,275 @@
+#include <postgres.h>
+
+#include "product_quantization.h"
+
+#include <assert.h>
+#include <float.h>
+#include <string.h>
+#include <time.h>
+#include <utils/builtins.h>
+
+#include "usearch.h"
+
+/*
+ * Get random index which is not already used
+ * This will be used to pick random centroids from the dataset
+ * When initializing clusters
+ * */
+static int get_available_index(int max_idx, int *used_indexes, int *used_indexes_size)
+{
+    int  idx;
+    int  i;
+    bool used = false;
+    while(true) {
+        idx = rand() % (max_idx + 1);
+        used = false;
+
+        for(i = 0; i < *used_indexes_size; i++) {
+            if(used_indexes[ i ] == idx) {
+                used = true;
+                break;
+            }
+        }
+
+        if(!used) break;
+    }
+    used_indexes[ *used_indexes_size ] = idx;
```

Comment:
who sizes `used_indexes` to make sure that this does not go out of bounds?

It seems you only want to write to `used_indexes` and `used_indexes_size` to append a single value to the end of the array that is the same value as the returned `idx` right?
if so, might be cleaner to make `used_indexes` and `used_indexes_size` `const` within this function, and update the arrays at callsite